Foundation/NSLock.swift

```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
     
     open var name: String?
 }
+
+private func timeSpecFrom(date: Date) -> timespec? {
+    let interval = date.timeIntervalSince1970
+    guard interval >= 0 else {
+        return nil
+    }
+
+    var ts = timespec()
+    ts.tv_sec = Int(interval.rounded(.down))
+    ts.tv_nsec = Int((interval - Double(ts.tv_sec)) * 1_000_000_000.0)
+    return ts
+}
+
+#if os(OSX) || os(iOS)
+
+private func allocateTimedLockData() -> UnsafeMutablePointer<_timed_lock_data>? {
+    let tld = UnsafeMutablePointer<_timed_lock_data>.allocate(capacity: 1)
+    if pthread_mutex_init(_CFGetTimedLockMutex(tld), nil) == 0 {
+        if pthread_cond_init(_CFGetTimedLockCond(tld), nil) == 0 {
+            return tld
+        } else {
+            pthread_mutex_destroy(_CFGetTimedLockMutex(tld))
+        }
+    }
+    tld.deallocate(capacity: 1)
+    return nil
+}
+
+private func deallocateTimedLockData(_ tld: UnsafeMutablePointer<_timed_lock_data>?) {
+    if let tld = tld {
+        pthread_mutex_destroy(_CFGetTimedLockMutex(tld))
+        pthread_cond_destroy(_CFGetTimedLockCond(tld))
+        tld.deallocate(capacity: 1)
+    }
+}
+
+// Emulate pthread_mutex_timedlock using pthread_cond_timedwait.
+// lock(before:) has obtained a shared pointer to a _time_lock_data structure
+// and the condition variable to be waited on.
+// unlock() will check for the existance of this structure and if its pointer is
+// non-nil, it will use pthread_cond_broadcast() to stop any waits in progress.
+private func timedLock(mutex: _PthreadMutexPointer,
+                       endTime: inout timespec,
+                       using tld: UnsafeMutablePointer<_timed_lock_data>) -> Bool {
+    pthread_mutex_lock(_CFGetTimedLockMutex(tld))
+    let status = pthread_cond_timedwait(_CFGetTimedLockCond(tld),
+                                        _CFGetTimedLockMutex(tld),
+                                        &endTime)
+    pthread_mutex_unlock(_CFGetTimedLockMutex(tld))
+
+    if status == 0 {
+        return pthread_mutex_trylock(mutex) == 0
```

Comment:
why are we trying to re-lock here?

Reply:
This is how the lock is actually acquired. The `pthread_mutex_lock` / `pthread_cond_timedwait` / `pthread_mutex_unlock` merely handles the waiting/timeout, the `pthread_mutex_trylock` is the actual (non-waiting) lock.

Reply:
@spevans sorry, now I'm completely lost, what is a non-waiting lock? And why do we need two locks for a timed lock?

Reply:
On macOS there are only two ways to lock: `pthread_mutex_lock` which waits indefinitely and `pthread_mutex_trylock` which tries to lock and returns immediately whether it locked or not (i.e. non-waiting).

`pthread_cond_timewait` uses a `pthread_mutex_t`/`pthread_cond_t` pair to perform its actions. So there are actually 2 mutexes and one condition variable involved here.

`mutex` is the actual lock that is trying to be acquired. The `_timed_lock_data` struct holds the `pthread_mutex_t`/`pthread_cond_t` pair for the `pthread_cond_timewait`.
Does that make sense?

Reply:
@spevans apologies, I didn't quite read the code correctly, thought this is all the same mutex :P.

Ok, in that case I still think it's not quite right as the `pthread_mutex_unlock(_CFGetTimedLockMutex(tld))` is done too early. If I'm not mistaken what could happen is the following: 

- thread 1 progresses until line  354 with `status == 0`
- thread 1 is now scheduled off CPU
- thread 2 arrives and also progresses until line 354 with `status == 0`
- thread 2 runs the function to completion and gets the lock `mutex` (with `pthread_mutex_trylock(mutex) == 0`)
- now thread 1 gets scheduled again and also tries to get the lock (with `pthread_mutex_trylock(mutex) == 0`) but because it's already taken by thread 2, it doesn't get it

now, in thread 1 `timedLock(...)` returns `false` even though the timeout might not have expired.


The same problem can happen if `pthread_cond_timedwait` spuriously wakes up. It's allowed to do that. Basically conditional variables are hardly ever used correctly if they're not in a loop. See [here](https://linux.die.net/man/3/pthread_cond_wait)

> Spurious wakeups from the pthread_cond_timedwait() or pthread_cond_wait() functions may occur.

please let me  know if I missed anything again and sorry agaiin

Reply:
@weissi Yes you are correct in that I have made a small error assuming that as soon as one thread gets the lock the others should return `false`, when they should actually retry the loop until the `limit: Date` is reached. Adding a loop would fix this and also work around any spurious wake ups.

I will add a fix for the above, however note that this solution is only used on macOS and the manual page for `pthread_cond_timedwait` doesn't actually mention spurious wakeups, it will only return on an invalid argument or a timeout.

On Linux, all of this is replaced by a single call of `pthread_mutex_timedlock`.